third_party/lua-cjson/lua_cjson.c

```diff
@@ -423,10 +439,13 @@ static int json_encode(lua_State *l) {
     struct luaL_serializer *cfg = luaL_checkserializer(l);
 
     if (lua_gettop(l) == 2) {
-        struct luaL_serializer user_cfg = *cfg;
+        struct luaL_serializer user_cfg;
+        luaL_serializer_create(&user_cfg);
+        luaL_serializer_copy_options(&user_cfg, cfg);
         luaL_serializer_parse_options(l, &user_cfg);
         lua_pop(l, 1);
         json_append_data(l, &user_cfg, 0, &encode_buf);
+        luaL_serializer_options_delete(&user_cfg);
```

Comment:
Sorry for the late comment, but JSON encoding (as well as decoding) can raise a Lua error. It seems, in this case we can get a memory leak.

As a simple fixup I can propose to allocate this serializer object as a userdata Lua object and keep it on the Lua stack.

Reply:
Thanks for pointing that out! I've fixed it by creating the `user_cfg` on Lua stack with `__gc` method, and poping it right away, so that the GC will clean everything up even in case encoding/decoding fails. I've also added a test for that, that will fail under ASAN test.

Reply:
We should keep it on the stack till the end of the function execution. Otherwise it may be collected on some Lua-land allocation (for example, `lua_newtable` or `lua_pushstring`). Let's push it on the 1st Lua stack slot (change with the value to encode using `lua_insert(L, -2);`).

Reply:
I've fixed it) there were some problems with it, because some C functions in the trace were calling `lua_` functions and were counting the stack positions from the bottom. I've also added some test that will fail under ASAN with the previous approach.

Reply:
Now it looks valid, thanks!